app/javascript/count.js

```diff
@@ -1,7 +1,15 @@
+const moment = require('moment')
+
 const resolutionMS = 1000
 
-document.addEventListener('turbolinks:load', () => {
-  Array.from(document.querySelectorAll('[data-ms]')).forEach(el => {
-    window.setTimeout(() => el.dataset.ms += resolutionMS, resolutionMS)
-  })
+// We attach to load instead of turbolinks:load because this function re-finds DOM elements every tick, so we can re-use
+// the same function. We choose to re-find DOM elements every tick because ActionCable can replace our element with a
+// new one rendered server-side at any moment.
+window.addEventListener('load', () => {
+  setInterval(() => {
```

Comment:
How accurate do we want the counting time to be?  `setInterval` doesn't guarantee the function be ran every X milliseconds, just that it will be run around that interval length.  When I was doing SRR stuff I found that even as high has 200-300ms intervals you can see the seconds be of variable length, and closer to 1000 I saw it actually skip numbers before.

Reply:
Oh wow, TIL! 🌠 How did you solve this in SRR? Should I implement something like this? https://medium.com/@swainson/accurate-javascript-timers-f71e0af5df32

Reply:
With the race start time as a data attribute I would use the timesync instance to calculate a duration.  That way it didn't rely on the timer being accurate since it was working with static times.  That link seems like it would accomplish things as well just in a different way, so whatever you think is easiest.

Reply:
I'd argue you just want 1s to be an integer multiple of the interval you choose here. And it needs to be lower than 1s to avoid "aliasing" / skipping. So 0.5s, 0.25s or 0.2s (avoid the last one due to floating point inaccuracies maybe) should be fine.

Reply:
Okay just to spell this out for my own understanding, it seems like there are two issues with setInterval:

**Precision**: an interval of `n` ms might trigger after `n +/- c` ms instead. At its best this makes the timer tick `c` ms early/late, which may or may not be noticeable by a human. At its worst this can cause skipping, e.g. `00:00:00` ticks first at 999 ms, then at 1002 ms, which jumps the timer from `00:00:00` to `00:00:02`.

**Drift**: Bad precision over long periods of time (and/or other issues?) can make the timer drift from the actual elapsed time. e.g. 4 hours in your timer might say `04:32:10` when the actual elapsed time is `04:31:56`. Even just a 1ms error every second for 4 hours adds up to 14 seconds off.
 
Having 1s be a multiple of our interval of is a partway solution from what I can see. It avoids skipping, but it's still prone to drift and the first issue with precision -- the length of a second can vary by an amount *up to the interval duration*.

```
|--one real second--|
|-------------------|-------------------|
:   :   :   :   :   :   :   :   :   :   : <- where ticks should happen (200ms)
^   ^   ^   ^   ^  ^    ^   ^   ^   ^   ^ <- where ticks do happen (200ms +/- c)
\-----------------------/
 this "second" is 1200ms
                        \---------------/
                        this one is 800ms
```

The weird part is that the smaller the tick interval, the *higher the chance* of this happening. A real-world second that elapses in the mid-point between two ticks is relatively safe from falling off the edge of a tick due to variance. As the gap between ticks shrinks, the size of that "safe zone" shrinks. But the tick interval is still the upper bound for how off a second can be, so smaller is still probably better.

`timesync` solves drift, but I'm not sure whether it solves precision. It seems like the length of a displayed second might still vary due to setInterval.

I just found [dbkaplun/driftless][1] which recursively re-schedules ticks the closer they get to occurring to be more accurate. I'll mess around with this and `timesync` and with the function I linked before and see what works. `timesync` seems 👌👌 tho.

[1]: https://github.com/dbkaplun/driftless

Reply:
@CryZe would having it be a multiple really help though since it's going to not only take time to execute the function, but could be interrupted by something else taking over causing it to not be ran precisely every multiple of the second?

@glacials Yeah, timesync solves the drift problem, but not the precision problem.  Driftless seems pretty cool, and it could solve the precision problem as long as stuff isn't happening earlier than it should be otherwise we're going to have the exact same problem where things just miss the new second and causes the next second to become "longer".